File: haskell/bob/Bob.hs
module Bob(responseFor) where

import Data.Char

responseFor :: String -> String
responseFor s
  | isSilentTreatment s = "Fine. Be that way!"
  | isShout s = "Woah, chill out!"
  | isQuestion s = "Sure."
  | otherwise = "Whatever."

isSilentTreatment :: String -> Bool
isSilentTreatment s =  all isSpace s

isQuestion :: String -> Bool
isQuestion s = last s == '?'

isShout :: String -> Bool
isShout s = any isAlpha s && not (any isLower s)

File: javascript/bob/bob.js
var Bob = function() {
  var isSilentTreatment = function(input) {
    return input.length === 0 || /^\s*$/.test(input);
  };

  var isShouting = function(input) {
    return input.toUpperCase() === input && input.toLowerCase() !== input;
  };

  var isQuestion = function(input) {
    return input.charAt(input.length - 1) === '?';
  };

  this.hey = function(input) {
    if (isSilentTreatment(input)) {
      return 'Fine. Be that way!';
    }

    if (isShouting(input)) {
      return 'Whoa, chill out!';
    }

    if (isQuestion(input)) {
      return 'Sure.';
    }

    return "Whatever.";
  };
};

module.exports = Bob;
